pkg/operator/etcd/pod.go

```diff
@@ -311,17 +312,31 @@ func (spec *Spec) Args() []string {
 	hostname := PodName(spec.LockserverName, spec.Index)
 	subdomain := PeerServiceName(spec.LockserverName)
 
-	listenPeerURLs := fmt.Sprintf("http://0.0.0.0:%d", peerPortNumber)
-	listenClientURLs := fmt.Sprintf("http://0.0.0.0:%d", clientPortNumber)
-	advertiseClientURLs := fmt.Sprintf("http://%s.%s:%d", hostname, subdomain, clientPortNumber)
+	listenPeerURLs := fmt.Sprintf("http://0.0.0.0:%d", PeerPortNumber)
+	listenClientURLs := fmt.Sprintf("http://0.0.0.0:%d", ClientPortNumber)
+	advertiseClientURLs := fmt.Sprintf("http://%s.%s:%d", hostname, subdomain, ClientPortNumber)
 
 	// Use static bootstrapping.
 	initialClusterToken := spec.LockserverName
-	initialAdvertisePeerURLs := fmt.Sprintf("http://%s.%s:%d", hostname, subdomain, peerPortNumber)
+	advertisePeerURLs := spec.AdvertisePeerURLs
+
+	// If peer URLs were not explicitly specified, generate them.
+	if len(advertisePeerURLs) != NumReplicas {
+		advertisePeerURLs = make([]string, 0, NumReplicas)
+		for i := 1; i <= NumReplicas; i++ {
+			peerName := PodName(spec.LockserverName, i)
+			advertisePeerURLs = append(advertisePeerURLs, fmt.Sprintf("http://%s.%s:%d", peerName, subdomain, PeerPortNumber))
+		}
+	}
+
+	// Set the address that this peer will advertise for itself.
+	initialAdvertisePeerURLs := advertisePeerURLs[spec.Index-1]
+
+	// Create list of peer addresses.
 	initialCluster := make([]string, 0, NumReplicas)
 	for i := 1; i <= NumReplicas; i++ {
```

Comment:
If possible with manual loops, I think it's better to index from 0. In this case we use i twice, and the second time we index at `i-1`. I personally would prefer PodName be with i+1 (which is clearer anyways, since it's not *really* the index), and index into advertisePeerURLs with just `i`.

Reply:
Done.